File: Cloud Storage Samples/Box Browser/cpp/box.cpp
/*
 * Cloud Storage 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudstorage.h"
#define LINE_LEN 100

class MyBox : public Box
{
public:
	
	virtual int FireSSLServerAuthentication(BoxSSLServerAuthenticationEventParams *e)
	{
		e->Accept = true;
		return 0;
	}
};

MyBox box1;

void printOptions()
{
	printf( "Type ? to select this menu, a number 1-5 to select the following choices:\n"
	        "1.  List Documents\n"
	        "2.  Download Document\n"
	        "3.  Upload Document\n"
	        "4.  Delete Document\n"
	        "5.  Quit\n");
}


char* resolveResourceType(int type) {
	switch (type)
	{
	case BRT_FILE: return "File";
	case BRT_FOLDER: return "Folder";
	default: return "Unknown";
	}
}


void displayDocuments(){

	int ret_code = box1.ListResources("");
	if(ret_code)
		goto error;

	printf("%-4.1s%-42.40s%-22.20s%\n\n","#","Title","Type");

	for(int i=0;i<box1.GetResourceCount();i++)
	{
		printf("%-4.1d%-42.40s%-22.20s%\n",i,box1.GetResourceName(i),resolveResourceType(box1.GetResourceType(i)));
	}
	return;

error:
	printf("\nError: %d", ret_code);
	if (box1.GetLastError())
	{
		printf( " \"%s\"\n", box1.GetLastError() );
	}
	printf("Press any key to continue");
	fflush(stdin);
	getchar();
	exit(ret_code);
}




int main(int argc, char **argv)
{
  char buffer[LINE_LEN+1];
  int ret_code = 0;

  printf("***************************************************************\n");
  printf("* This demo shows how to use the Box component to list,       *\n");
  printf("* download, upload, and delete documents. To begin, you will  *\n");
  printf("* first need to setup an app in the box developer console     *\n");
  printf("* to obtain your App Key and App Secret. You will also need   *\n");
  printf("* to setup a Redirect URI to http://localhost:PORT, where     *\n");
  printf("* PORT is the port number configured in the code.             *\n");
  printf("***************************************************************\n\n");

  printf("Client ID: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  box1.SetOAuthClientId(buffer);
  
  
  printf("Client Secret: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  box1.SetOAuthClientSecret(buffer);
 
  
  box1.SetOAuthServerAuthURL("https://app.box.com/api/oauth2/authorize");
  box1.SetOAuthServerTokenURL("https://api.box.com/oauth2/token");
  box1.Config("OAuthWebServerPort=7777");

  printf("\nPerforming OAuth authentication ...\n");

  box1.Authorize();

  ret_code = box1.GetLastErrorCode();

  if(ret_code)
    goto error;

	printf("\n");
	printOptions();

	while (1)
	{
		printf( "\n> " );
		scanf("%80s",buffer);

		if ( ! strcmp(buffer, "?") )
		{
			printOptions();
		}

		else if ( ! strcmp(buffer, "1") ) //List documents
		{
			displayDocuments();
		}

		else if ( ! strcmp(buffer, "2") ) //Download document
		{
			printf("\nDocument Number: ");
			scanf("%80s",buffer);
      int docNum = atoi(buffer);
			
			printf("Download Directory: ");
			scanf("%80s",buffer);
			strcat(buffer,"\\");
			strcat(buffer,box1.GetResourceName(docNum));
			box1.SetLocalFile(buffer);

			ret_code = box1.DownloadFile(box1.GetResourceId(docNum)); //Use the default file format
			if(ret_code)
				goto error;

			printf("\n Download Successful\n");
		}

		else if ( ! strcmp(buffer, "3") ) //Upload document
		{
			printf("\nLocal File: ");
			scanf("%80s",buffer);

			box1.SetLocalFile(buffer);
			
			printf("Name of New File on Server: ");
			scanf("%80s",buffer);
      		box1.UploadFile(buffer,"");
			ret_code = box1.GetLastErrorCode();

			if(ret_code)
				goto error;

			displayDocuments();
		}

		else if ( ! strcmp(buffer, "4") ) //Delete Document
		{
			printf("\nDocument Number: ");
			scanf("%80s",buffer);

			ret_code = box1.DeleteResource(box1.GetResourceId(atoi(buffer)));

			if(ret_code)
				goto error;

			displayDocuments();
		}

		else if ( ! strcmp(buffer, "5") )
		{
			exit(0);
		}

		else if ( ! strcmp(buffer, "") )
		{
			// Do nothing
		}
		else
		{
			printf( "Please select a number from [1-5].\n" );
		} // end of command checking
	}  // end of main while loop
	return ret_code;

error:
	printf("\nError: %d", ret_code);
	if (box1.GetLastErrorCode())
	{
		printf( " \"%s\"\n", box1.GetLastError() );
	}
	printf("Press any key to continue");
	fflush(stdin);
	getchar();
	return ret_code;

}

File: Cloud Storage Samples/Box Browser/java/box.java
/*
 * Cloud Storage 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import cloudstorage.*;

public class box extends ConsoleDemo {
	Box box1 = new Box();

	public box() {
		String buffer = "";
		try {
			box1.addBoxEventListener(new cloudstorage.DefaultBoxEventListener() {
						public void SSLServerAuthentication(
								BoxSSLServerAuthenticationEvent e) {
							e.accept = true;
						}
					});

			System.out.println("********************************************************************************");
			System.out.println("This demo shows how to List, Upload, Download, and Delete documents from Box    ");
			System.out.println("To begin, you will first need to setup an app in the Box developer console      ");
			System.out.println("to obtain your App Key and App Secret. You will also need to setup a Redirect   ");
			System.out.println("URI to http://localhost:PORT, where PORT is the port number configured below.   ");
			System.out.println("********************************************************************************");
			
			box1.getOAuth().setClientId(prompt("Client Id", ":"));
			box1.getOAuth().setClientSecret(prompt("Client Secret", ":"));
			box1.getOAuth().setServerAuthURL("https://app.box.com/api/oauth2/authorize");
			box1.getOAuth().setServerTokenURL("https://api.box.com/oauth2/token");
			box1.config("OAuthWebServerPort=" + Integer.parseInt(prompt("Redirect URI Port", ":")));
			box1.authorize();
			
			printMenu();
			while (!buffer.equals("q")) {
				System.out.print("Enter command: ");
				buffer = input();
				if (buffer.toLowerCase().equals("ls")) {
					ListDocuments();
				} else if (buffer.toLowerCase().equals("get")) {
					System.out.print("Document #: ");
					BoxResource r = box1.getResources().get(Integer.valueOf(input()));
					System.out.print("Local Directory: ");
					
					box1.setLocalFile(input() + "\\" + r.getName());
					box1.downloadFile(r.getId()); //Use the default file format
					
					System.out.println("\nDownload Complete.\n");
				} else if (buffer.toLowerCase().equals("put")) {
					System.out.print("Local File: ");
					box1.setLocalFile(input());
					File mFile = new File(box1.getLocalFile());
					box1.uploadFile(mFile.getName(), "");
					
					ListDocuments();
				} else if (buffer.toLowerCase().equals("del")) {
					System.out.print("Document #: ");
					BoxResource r = box1.getResources().get(Integer.valueOf(input()));
					
					box1.deleteResource(r.getId());
					
					ListDocuments();
				} else if (buffer.toLowerCase().equals("q")) {
					System.exit(0);
				} else if (buffer.toLowerCase().equals("?")) {
					printMenu();
				}
			}
		} catch (CloudStorageException e) {
			System.out.println(e.getMessage());
			System.exit(e.getCode());
			return;
		} catch (Exception e) {
			System.out.println(e.getMessage());
		}
	}

	private void printMenu() {
		System.out.println("\n?\t-\tHelp\n" + "ls\t-\tList Documents\n"
				+ "get\t-\tDownload Document\n" + "put\t-\tUpload Document\n"
				+ "del\t-\tDelete Document\n" + "q\t-\tQuit\n");
	}

	private void ListDocuments() {
		try {
			box1.listResources("");

			System.out.print("\n");
			System.out.format("%1$-3s%2$-40s%3$-20s%4$-30s\n","#","Title","Type","Id");
			System.out.println("--------------------------------------------------------------------------------");
			
			for (int i = 0; i < box1.getResources().size(); i++) {
				System.out.format("%1$-3s%2$-40.18s%3$-20s%4$-30s\n", String.valueOf(i), box1.getResources().get(i).getName(),
					ResolveResourceType(box1.getResources().get(i).getType()), box1.getResources().get(i).getId());
			}
			
			System.out.print("\n");
			
		} catch (CloudStorageException e) {
			System.out.println(e.getMessage());
			System.exit(e.getCode());
			return;
		} catch (Exception e) {
			System.out.println(e.getMessage());
		}
	}

	private String ResolveResourceType(int type) {
		switch (type) {
		case BoxResource.brtFile:	return "File";
		case BoxResource.brtFolder: return "Folder";
		default: return "Unknown";
		}
	}
	
	public static void main(String[] args) {
		try {
			new box();
		} catch (Exception ex) {
			System.out.println("Exception: " + ex.getMessage());
		}
	}

}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}



